content/browser/attribution_reporting/attribution_storage_sql.cc

```diff
@@ -2999,6 +3388,272 @@ bool AttributionStorageSql::StoreAttributionReport(AttributionReport& report) {
   return true;
 }
 
+AggregatableResult AttributionStorageSql::PayAllOrNothing(
+          attribution_reporting::AttributionWindow attribution_window,
+          const attribution_reporting::SuitableOrigin& querying_origin,
+          double required_budget) {
+
+  std::vector<uint64_t> attribution_epochs;
+  for (uint64_t i=attribution_window.epoch_start(); 
+          i<=attribution_window.epoch_end(); ++i) {
+    attribution_epochs.push_back(i);
+  }
+  return PayAllOrNothing(attribution_epochs, querying_origin, required_budget);
+}
+
+AggregatableResult AttributionStorageSql::PayAllOrNothing(
+          std::vector<uint64_t> attribution_epochs,
+          const attribution_reporting::SuitableOrigin& querying_origin,
+          double required_budget) {
+
+  // Starting atomic operation
+  sql::Transaction transaction(&db_);
+  if (!transaction.Begin()) {
+    return AggregatableResult::kInternalError;
+  }
+
+  // Obtain remaining budget per epoch for all attribution epochs  
+  std::ostringstream oss; 
+  oss << \
+    "SELECT F.epoch, (F.initial_budget - F.consumed_budget) AS remaining_budgets "
+      "FROM per_origin_filters F "
+      "WHERE F.origin=? "
+      "AND F.epoch IN (";
+
+  auto& items = attribution_epochs;
+  if (!items.empty()) {
+    std::copy(items.begin(), items.end() - 1, std::ostream_iterator<uint64_t>(oss, ", "));
+    oss << items.back();
+  }
+  oss << ")";
+
+  sql::Statement statement(db_.GetUniqueStatement(oss.str().c_str()));
+  statement.BindString(0, net::SchemefulSite(querying_origin).Serialize());
+
+  base::flat_map<uint64_t, double> remaining_budgets;
+  while (statement.Step()) {
+      remaining_budgets[statement.ColumnInt64(0)] = statement.ColumnDouble(1);
+  }
+
+  // For epochs not already stored in the DB initialize their
+  // remaining budget and mark them to be inserted now
+  std::vector<uint64_t> epochs_to_insert;
+  std::vector<uint64_t> epochs_to_update;
+  for (auto& epoch : attribution_epochs) {
+    auto it = remaining_budgets.find(epoch);
+    if (it == remaining_budgets.end()) {
+      epochs_to_insert.push_back(epoch);
+      remaining_budgets[epoch] = kInitialBudget;
+    } else {
+      epochs_to_update.push_back(epoch);
+    }
+  }
+  
+  // -------------------------------------------------------
+  LOG(INFO) << "Epochs to insert:";
+  for (auto epoch : epochs_to_insert) {
+    LOG(INFO) << epoch;
+  }
+  LOG(INFO) << "Epochs to update:";
+  for (auto epoch : epochs_to_update) {
+    LOG(INFO) << epoch;
+  }
+  
+  LOG(INFO) << "Remaining Budgets:";
+  for (const auto& pair : remaining_budgets) {
+    LOG(INFO) << "- Epoch " << pair.first << " Budget " << pair.second;      
+  }
+// -------------------------------------------------------
+
+  // Check if all remaining budgets are enough for required_budget
+  for (const auto& pair : remaining_budgets) {
+    if (pair.second < required_budget) {
+      return AggregatableResult::kInsufficientBudget;
+    }
+  }
+
```

Comment:
Is this where Optimization 4 is missing?

Reply:
I think incorporating the opt 4 is trickier in general. Because I have already computed the whole partition-report in advance it is not easy to null specific epoch-reports upon seeing that the epochs don't have enough budget. Checking budgets happens later in the execution. My solution will be to delete impressions from depleted epochs instead. If I do that optimization 4 becomes optimization 3.